### ISSUE TITLE
unauthenticated vs authenticated user id confusion

### ISSUE THREAD

**Issue by corehack** [Pylons/pyramid#3014]
I was initially confused why `request.user` recipe is using unauthenticated user id instead the authenticated one. And seems like i'm not the only one:
https://stackoverflow.com/questions/12547646/pyramid-security-is-getting-user-info-from-a-database-with-unauthenticated-user

Now I see there's argument that authenticated_userid is potentially costly. I understand that. But why not make `authenticated_userid` reified attribute and make it cheap ?
OTOH, using `unauthenticated_userid` in the recipe can create bugs in (newbie) developer's code, if he will just assume that `request.user` is always valid user for the current request (while its not). Also, unauthenticated_user for me as developer does not matter at all outside of the authentication process (it is a "private variable" of the authentication process).
Am I missing any edge cases where authentication has to be done more than one time per request ? Seems highly unlikely for me.

In summary, my suggestions:
1. Always use authenticated_user as a source for the user object
2. Make authenticated_userid reified or in other words lazy cached per-request attribute to make it cheap for everyday use

**Comment by mmerickel** [comment]
I updated the stackoverflow question with the correct response[1] which I will paste here for completeness.

> It has nothing to do with performance.
>
>The reason that it uses `unauthenticated_userid` is because it makes it easier to reuse the authentication policy between applications with smaller modifications required. Your application needs a "source of truth" for whether the user is allowed to be considered authenticated and usually the policy's internal logic is not enough to make this determination. A valid cookie is nice, but you usually want to verify it with your backend before trusting it. Great, so where do we put that logic? Well `unauthenticated_userid` doesn't make sense because it is the reusable part of the policy that focuses specifically on parsing the request headers. You could put it into `authenticated_userid` but this method is not the one you normally care about in your application. You normally use `request.user` in your apps (rarely do you probably care about `request.authenticated_userid` directly) and lastly the `request.user` is a superset of functionality - it provides an entire user object, not just an id. It would be silly to verify the id without verifying the entire object in most cases. We can only have one "source of truth" and so the recipe declares it to be `request.user`. The groupfinder (and thus `authenticated_userid`) can now depend on `request.user` and trust that what it gets back from there has been properly verified with the backend. Also `request.user` is already reified and thus speeds up subsequent calls to `request.authenticated_userid` naturally.

If you would like to submit any updates to the docs to indicate what's going on here they would be appreciated but hopefully it's clear what is going on now.

[1] https://stackoverflow.com/questions/12547646/pyramid-security-is-getting-user-info-from-a-database-with-unauthenticated-user/43590930#43590930
